README.md
<p align="left">
  <a href="https://pub.dev/packages/flutter_facebook_auth"><img alt="pub version" src="https://img.shields.io/pub/v/flutter_facebook_auth?color=%2300b0ff&label=flutter_facebook_auth&style=flat-square"></a>
  
  <img alt="last commit" src="https://img.shields.io/github/last-commit/the-meedu-app/flutter-facebook-auth?color=%23ffa000&style=flat-square"/> 
  <img alt="license" src="https://img.shields.io/github/license/the-meedu-app/flutter-facebook-auth?style=flat-square"/>
  <img alt="stars" src="https://img.shields.io/github/stars/the-meedu-app/flutter-facebook-auth?style=social"/>
</p>

---

# flutter_facebook_auth

Flutter plugin to make easy the facebook authentication in your flutter app. iOS and Android are supported.

## **install on Android**

Go to https://developers.facebook.com/docs/facebook-login/android/?locale=en and read the next documentation.

- Edit Your Resources and Manifest
- Associate Your Package Name and Default Class with Your App
- Provide the Development and Release Key Hashes for Your App

---

## **Install on iOS**

In your Podfile uncomment the next line (You need set the minimum target to 9.0 or higher)

```
platform :ios, '9.0'
```

Configure `Info.plist`

```xml
<key>CFBundleURLTypes</key>
<array>
  <dict>
    <key>CFBundleURLSchemes</key>
    <array>
      <string>fb{your-app-id}</string>
    </array>
  </dict>
</array>
<key>FacebookAppID</key>
<string>{your-app-id}</string>
<key>FacebookDisplayName</key>
<string>{your-app-name}</string>
<key>LSApplicationQueriesSchemes</key>
<array>
  <string>fbapi</string>
  <string>fb-messenger-share-api</string>
  <string>fbauth2</string>
  <string>fbshareextension</string>
</array>
```

## **On iOS you need Swift support**

The plugin is written in `Swift`, so your project needs to have Swift support enabled. If you've created the project using `flutter create -i swift [projectName]` you are all set. If not, you can enable Swift support by opening the project with XCode, then choose `File -> New -> File -> Swift File`. XCode will ask you if you wish to create Bridging Header, click yes.

---

## If you have implement another providers (Like Google) in your app you should merge values in Info.plist

Check if you already have CFBundleURLTypes or LSApplicationQueriesSchemes keys in your Info.plist. If you have, you should merge their values, instead of adding a duplicate key.

Example with Google and Facebook implemetation:

```xml
<key>CFBundleURLTypes</key>
<array>
  <dict>
    <key>CFBundleTypeRole</key>
    <string>Editor</string>
    <key>CFBundleURLSchemes</key>
    <array>
      <string>fb{your-app-id}</string>
      <string>com.googleusercontent.apps.{your-app-specific-url}</string>
    </array>
  </dict>
</array>
```

---

## NOTE

for Objective-C projects (correctly works is not granted because this plugin was written with swift)

---

### **METHODS**

Just use `FacebookAuth.instance`. NOTE: all methods are **asynchronous**.

- `.login({List<String> permissions = const ['email', 'public_profile'] })` : request login with a list of permissions.

  The `public_profile` permission allows you read the next fields
  `id, first_name, last_name, middle_name, name, name_format, picture, short_name`

  For more info go to https://developers.facebook.com/docs/facebook-login/permissions/

  return one instance of `LoginResult` class:

  ```
  {
    status: 200,
    accessToken: {
      expires: 1573493493209,
      declinedPermissions: [],
      grantedPermissions: [public_profile, email],
      userId: 3003332493073668,
      token: EAAVDwBAE5lndhpg17DHFZABzh6QKiAZC42Qljcub9gib52L5CPEXvhk2ZBEa7LlOuyastmmkZBfwP7dKW6Xi4tvrTw8DToO2M2kMcau6CXsYtyys7WZAWV3XaMPnhuVauo5ghtGpnhJvZAtMKqlsgbV5GklPAYZD
    }
  }
  ```

  if `status` is 200 the login was successfull.

- `.logOut()` : close the current facebook session.

- `.isLogged` : check if the user has an active facebook session. The response will be `null` if the user is not logged.
  return one instance of `AccessToken` class:

  ```
  {
      expires: 1573493493209,
      userId: 3003332493073668,
      token: EAATaHWA7VDwBAE5lndhpg17DHFZABzh6QKiAZC42Qljcub9gib52L5CPEXvhk2ZBEa7LlOuytmmkZBfwP7dKW6Xi4XCO2M2kMcau6CXsYtyys7WZAWV3XaMPnhuVauo5ghtGpnhJvZAtMKqlsgbV5GklPAYZD,
      declinedPermissions: [],
      grantedPermissions: [public_profile, email],
    }
  ```

- `.getUserData({String fields = "name,email,picture"})` : get the user info only if the user is logged.

  Expected response:

  ```
  {
  email = "dsmr.apps@gmail.com";
  id = 3003332493073668;
  name = "Darwin Morocho";
  picture =     {
      data =         {
          height = 50;
          "is_silhouette" = 0;
          url = "https://platform-lookaside.fbsbx.com/platform/profilepic/?asid=3003332493073668&height=50&width=50&ext=1570917120&hash=AeQMSBD5s4QdgLoh";
          width = 50;
      };
  };
  }
  ```

### **EXAMPLE**

```dart
import 'dart:convert';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_facebook_auth/flutter_facebook_auth.dart';

void main() => runApp(MyApp());

String prettyPrint(Map json) {
  JsonEncoder encoder = new JsonEncoder.withIndent('  ');
  String pretty = encoder.convert(json);
  return pretty;
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  dynamic _userData;
  AccessToken _accessToken;
  bool _checking = true;

  String twitterStatus = "No Logged";

  @override
  void initState() {
    super.initState();
    _checkIfIsLogged();
  }

  _printCredentials(LoginResult result) {
    _accessToken = result.accessToken;
    print("userId: ${result.accessToken.userId}");
    print("token: ${prettyPrint(_accessToken.toJson())}");
    print("expires: ${result.accessToken.expires}");
    print("grantedPermission: ${result.accessToken.grantedPermissions}");
    print("declinedPermissions: ${result.accessToken.declinedPermissions}");
  }

  _checkIfIsLogged() async {
    final AccessToken accessToken = await FacebookAuth.instance.isLogged;
    setState(() {
      _checking = false;
    });
    if (accessToken != null) {
      print("is Logged:::: ${prettyPrint(accessToken.toJson())}");
      // now you can call to  FacebookAuth.instance.getUserData();
      final userData = await FacebookAuth.instance.getUserData();
      // final userData = await FacebookAuth.instance.getUserData(fields:"email,birthday");
      _accessToken = accessToken;
      setState(() {
        _userData = userData;
      });
    }
  }

  _login() async {
    setState(() {
      _checking = true;
    });
    final result = await FacebookAuth.instance.login();
    // final result = await FacebookAuth.instance.login(permissions:['email','user_birthday']);

    _checking = false;
    switch (result.status) {
      case FacebookAuthLoginResponse.ok:
        _printCredentials(result);
        // get the user data
        final userData = await FacebookAuth.instance.getUserData();
        // final userData = await _fb.getUserData(fields:"email,birthday");
        _userData = userData;
        break;
      case FacebookAuthLoginResponse.cancelled:
        print("login cancelled");
        break;
      default:
        print("login failed");
    }

    setState(() {});
  }

  _logOut() async {
    await FacebookAuth.instance.logOut();
    _accessToken = null;
    setState(() {
      _userData = null;
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Facebook Auth Example'),
        ),
        body: _checking
            ? Center(
                child: CircularProgressIndicator(),
              )
            : SingleChildScrollView(
                child: Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.center,
                    children: <Widget>[
                      Text(
                        _userData != null
                            ? prettyPrint(_userData)
                            : "NO LOGGED",
                      ),
                      SizedBox(height: 20),
                      _accessToken != null
                          ? Text(
                              prettyPrint(_accessToken.toJson()),
                            )
                          : Container(),
                      SizedBox(height: 20),
                      CupertinoButton(
                        color: Colors.blue,
                        child: Text(
                          _userData != null ? "LOGOUT" : "LOGIN",
                          style: TextStyle(color: Colors.white),
                        ),
                        onPressed: _userData != null ? _logOut : _login,
                      ),
                      SizedBox(height: 50),
                    ],
                  ),
                ),
              ),
      ),
    );
  }
}
```

## **Using with firebase_auth**

```dart
import 'package:firebase_auth/firebase_auth.dart';
.
.
.
.


final FirebaseAuth _auth = FirebaseAuth.instance;
.
.
.

 // this line do auth in firebase with your facebook credential. Just pass your facebook token (String)
 final OAuthCredential credential =  FacebookAuthProvider.credential( _accessToken.token); // _token is your facebook access token as a string

// FirebaseUser is deprecated
final User user = (await _auth.signInWithCredential(credential)).user;
```


CHANGELOG.md
## 0.3.2
- flutter: ">=1.0.0 <2.0.0"

## 0.3.1+1
- Added badge version

## 0.3.1
- Updated FBSDKCoreKit and FBSDKLoginKit to ~> 7.1.0 on iOS

## 0.3.0
- Added granted and declined permissions on AccessToken class, and removed AppDelegate.swift
- Updated example
- Updated documentation

## 0.2.4
- Updated documentation and isLogged method to catch posible exceptions 

## 0.2.3
- Updated to Supporting the new Android plugins APIs 

## 0.2.2+3
- Removed dangerous log 

## 0.2.2+2
- Updated Docs to allow login with facebook and twitter in the same flutter app

## 0.2.2+1
- Updated Readme.md

## 0.2.2
- Added documentation for ApplicationDelegate.shared on iOS

## 0.2.1
- Removed dangerous log

## 0.2.0
- Updated Facebook Core version to remove deprecated UIWebView on iOS.
- Updated methods.


## 0.1.1
- iOS version









ios/Classes/SwiftFlutterFacebookAuthPlugin.swift
import Flutter
import UIKit

import FBSDKCoreKit
import FBSDKLoginKit

public class SwiftFlutterFacebookAuthPlugin: NSObject, FlutterPlugin {
    
    let loginManager : LoginManager = LoginManager()
    var pendingResult: FlutterResult? = nil
    
    
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "ec.dina/flutter_facebook_auth", binaryMessenger: registrar.messenger())
        let instance = SwiftFlutterFacebookAuthPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        registrar.addApplicationDelegate(instance)
    }
    
    /// START ALLOW HANDLE NATIVE FACEBOOK APP
    public func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [AnyHashable : Any] = [:]) -> Bool {
        var options = [UIApplication.LaunchOptionsKey: Any]()
        for (k, value) in launchOptions {
            let key = k as! UIApplication.LaunchOptionsKey
            options[key] = value
        }
        ApplicationDelegate.shared.application(application,didFinishLaunchingWithOptions: options)
        return true
    }
    
   public func application( _ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey : Any] = [:] ) -> Bool {
        let processed = ApplicationDelegate.shared.application(
            app, open: url,
            sourceApplication: options[UIApplication.OpenURLOptionsKey.sourceApplication] as? String,
            annotation: options[UIApplication.OpenURLOptionsKey.annotation])
        return processed;
    }
    /// END ALLOW HANDLE NATIVE FACEBOOK APP
    
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        
        let args = call.arguments as? [String: Any]
        
        switch call.method{
            
        case "login":
            let permissions = args?["permissions"] as! [String]
            self.login(permissions: permissions, flutterResult: result)
            
            
        case "isLogged":
            let isOK = setPendingResult(methodName: "isLogged", flutterResult: result)
            if(!isOK){
                return
            }
            if AccessToken.isCurrentAccessTokenActive {
                let accessToken = getAccessToken(accessToken: AccessToken.current!)
                finishWithResult(data: accessToken as Dictionary)
            }else{
                finishWithResult(data: nil)
            }
            
            
        case "getUserData":
            let fields = args?["fields"] as! String
            getUserData(fields: fields, flutterResult: result)
            
        case "logOut":
            let isOK = setPendingResult(methodName: "logOut", flutterResult: result)
            if(!isOK){
                return
            }
            loginManager.logOut()
            finishWithResult(data: nil)
            
            
            
        default:
            result(FlutterError(code: "404", message: "No such method", details: nil))
        }
        
    }
    
    
    
    private func login(permissions: [String], flutterResult: @escaping FlutterResult){
        
        let isOK = setPendingResult(methodName: "login", flutterResult: flutterResult)
        if(!isOK){
            return
        }
        
        let viewController: UIViewController = (UIApplication.shared.delegate?.window??.rootViewController)!
        
        loginManager.logIn(permissions: permissions, from: viewController, handler: { (result, error) -> Void in
            
            if(error==nil){
                let fbloginresult : LoginManagerLoginResult = result!
                
                if(fbloginresult.isCancelled) {
                    self.finishWithResult(data: ["status":403])
                } else {
                    
                  
                    
                    self.finishWithResult(data: [
                        "status":200,
                        "accessToken": self.getAccessToken(accessToken: fbloginresult.token!),
                        "grantedPermissions": Array(fbloginresult.grantedPermissions),
                        "declinedPermissions": Array(fbloginresult.declinedPermissions)
                    ])
                }
            } else{
                self.finishWithError(message: "error make sure that your  Info.plist is configured")
            }
            
        })
        
    }
    
    
    
    
    
    
    
    private func getUserData(fields: String, flutterResult: @escaping FlutterResult) {
        print("init getUserData:")
        
        let isOK = setPendingResult(methodName: "getUserData", flutterResult: flutterResult)
        if(!isOK){
            return
        }
        
        let graphRequest : GraphRequest = GraphRequest(graphPath: "me", parameters: ["fields":fields])
        graphRequest.start(completionHandler: { (connection, result, error) -> Void in
            if ((error) != nil) {
                print("error getUserData: \(String(describing: error))")
                self.finishWithError(message: "error get user data from facebook Graph, please check your fileds and your permissions")
            } else {
                let resultDic = result as! NSDictionary
                self.finishWithResult(data: resultDic)
            }
        })
    }
    
    
    private func setPendingResult(methodName: String, flutterResult: @escaping FlutterResult) -> Bool {
        if(pendingResult != nil){
            pendingResult!(FlutterError(code: "500", message: "\(methodName) called while another Facebook login operation was in progress.", details: nil))
            
            return false
        }
        pendingResult = flutterResult;
        return true
    }
    
    private func finishWithResult(data: Any?){
        if (pendingResult != nil) {
            pendingResult!(data)
            pendingResult = nil
        }
    }
    
    
    private func finishWithError(message: String){
        if (pendingResult != nil) {
            pendingResult!(FlutterError(code: "500", message: message, details: nil))
            pendingResult = nil
        }
    }
    
    
    
    
    private func getAccessToken(accessToken: AccessToken) -> [String : Any] {
        let data = [
            "token": accessToken.tokenString,
            "userId": accessToken.userID,
            "expires": Int64((accessToken.expirationDate.timeIntervalSince1970*1000).rounded()),
            "grantedPermissions":accessToken.permissions.map {item in item.name},
            "declinedPermissions":accessToken.declinedPermissions.map {item in item.name},
            ] as [String : Any]
        
        return data;
        
    }
    
    
}


example/ios/Flutter/flutter_export_environment.sh
#!/bin/sh
# This is a generated file; do not edit or check into version control.
export "FLUTTER_ROOT=/Users/darwin/development/flutter"
export "FLUTTER_APPLICATION_PATH=/Users/darwin/Documents/flutter/flutter-facebook-auth/example"
export "FLUTTER_TARGET=/Users/darwin/Documents/flutter/flutter-facebook-auth/example/lib/main.dart"
export "FLUTTER_BUILD_DIR=build"
export "SYMROOT=${SOURCE_ROOT}/../build/ios"
export "OTHER_LDFLAGS=$(inherited) -framework Flutter"
export "FLUTTER_FRAMEWORK_DIR=/Users/darwin/development/flutter/bin/cache/artifacts/engine/ios"
export "FLUTTER_BUILD_NAME=1.0.0"
export "FLUTTER_BUILD_NUMBER=1"
export "DART_DEFINES=flutter.inspector.structuredErrors%3Dtrue"
export "DART_OBFUSCATION=false"
export "TRACK_WIDGET_CREATION=true"
export "TREE_SHAKE_ICONS=false"
export "PACKAGE_CONFIG=.packages"


android/src/main/java/ec/dina/flutter_facebook_auth/FacebookAuth.java
package ec.dina.flutter_facebook_auth;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;

import com.facebook.AccessToken;
import com.facebook.CallbackManager;
import com.facebook.FacebookCallback;
import com.facebook.FacebookException;
import com.facebook.GraphRequest;
import com.facebook.GraphResponse;
import com.facebook.login.LoginManager;
import com.facebook.login.LoginResult;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;

import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.PluginRegistry;

public class FacebookAuth {


    private final CallbackManager callbackManager;
    private LoginManager loginManager;
    FacebookLoginResultDelegate resultDelegate;


    FacebookAuth() {
        loginManager = LoginManager.getInstance();
        callbackManager = CallbackManager.Factory.create();
        resultDelegate = new FacebookLoginResultDelegate(callbackManager);
        loginManager.registerCallback(callbackManager, resultDelegate);
    }


    public void login(Activity activity, List<String> permissions, MethodChannel.Result result) {
        boolean isOK = resultDelegate.setPendingResult("login", result);

        if (isOK) {
            loginManager.logIn(activity, permissions);
        }
    }


    public void isLogged(MethodChannel.Result result) {
        boolean isOK = resultDelegate.setPendingResult("isLogged", result);
        if (isOK) {
            if (AccessToken.isDataAccessActive()) {
                HashMap data = getAccessToken(AccessToken.getCurrentAccessToken());
                resultDelegate.finishWithResult(data);
            } else {
                resultDelegate.finishWithResult(null);
            }
        }


    }


    public void getUserData(String fields, final MethodChannel.Result result) {

        boolean isOK = resultDelegate.setPendingResult("getUserData", result);
        if (isOK) {
            GraphRequest request = GraphRequest.newMeRequest(
                    AccessToken.getCurrentAccessToken(),
                    new GraphRequest.GraphJSONObjectCallback() {
                        @Override
                        public void onCompleted(JSONObject object, GraphResponse response) {
                            try {
                                resultDelegate.finishWithResult(object.toString());
                            } catch (Exception e) {
                                resultDelegate.finishWithError(e.getMessage());
                            }

                        }
                    });
            Bundle parameters = new Bundle();
            parameters.putString("fields", fields);
            request.setParameters(parameters);
            request.executeAsync();
        }


    }


    public void logOut(MethodChannel.Result result) {

        boolean isOK = resultDelegate.setPendingResult("logOut", result);
        if (isOK) {
            loginManager.logOut();
            resultDelegate.finishWithResult(null);
        }

    }


    static HashMap getAccessToken(final AccessToken accessToken) {
        return new HashMap<String, Object>() {{
            put("token", accessToken.getToken());
            put("userId", accessToken.getUserId());
            put("expires", accessToken.getExpires().getTime());
            put("grantedPermissions", new ArrayList<>(accessToken.getPermissions()));
            put("declinedPermissions",  new ArrayList<>(accessToken.getDeclinedPermissions()));
        }};
    }


}


class FacebookLoginResultDelegate implements FacebookCallback<LoginResult>, PluginRegistry.ActivityResultListener {
    private final CallbackManager callbackManager;
    private MethodChannel.Result pendingResult;

    FacebookLoginResultDelegate(CallbackManager callbackManager) {
        this.callbackManager = callbackManager;
    }

    @Override
    public void onSuccess(LoginResult loginResult) {
        HashMap<String, Object> res = new HashMap<>();
        res.put("status", 200);
        res.put("grantedPermissions", new ArrayList<>(loginResult.getRecentlyGrantedPermissions()));
        res.put("declinedPermissions", new ArrayList<>(loginResult.getRecentlyDeniedPermissions()));
        res.put("accessToken", FacebookAuth.getAccessToken(loginResult.getAccessToken()));
        finishWithResult(res);
    }

    @Override
    public void onCancel() {
        HashMap<String, Object> res = new HashMap<>();
        res.put("status", 403);
        finishWithResult(res);
    }

    @Override
    public void onError(FacebookException error) {
        finishWithError(error.getMessage());
    }

    @Override
    public boolean onActivityResult(int requestCode, int resultCode, Intent data) {
        return callbackManager.onActivityResult(requestCode, resultCode, data);
    }

    boolean setPendingResult(String methodName, MethodChannel.Result result) {
        if (pendingResult != null) {
            result.error("500",
                    methodName + " called while another Facebook " +
                            "login operation was in progress.",
                    null
            );
            return false;
        }

        pendingResult = result;
        return true;

    }

    void finishWithResult(Object result) {
        if (pendingResult != null) {
            pendingResult.success(result);
            pendingResult = null;
        }
    }


    void finishWithError(String message) {
        if (pendingResult != null) {
            pendingResult.error("500", message, null);
            pendingResult = null;
        }
    }

}

android/src/main/java/ec/dina/flutter_facebook_auth/FlutterFacebookAuthPlugin.java
package ec.dina.flutter_facebook_auth;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.util.Log;

import androidx.annotation.NonNull;

import java.util.List;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;


/**
 * FlutterFacebookAuthPlugin
 */
public class FlutterFacebookAuthPlugin implements MethodCallHandler, FlutterPlugin, ActivityAware {

    private static final String CHANNEL_NAME = "ec.dina/flutter_facebook_auth";


    private FacebookAuth facebookAuth = new FacebookAuth();
    private ActivityPluginBinding activityPluginBinding;

    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private MethodChannel channel;


    /**
     * Plugin registration.
     */
    public static void registerWith(Registrar registrar) {
        Log.i("facebook:", "registerWith");
        FlutterFacebookAuthPlugin plugin = new FlutterFacebookAuthPlugin();
        plugin.onAttachedToEngine(registrar.messenger());
    }

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
        this.onAttachedToEngine(binding.getBinaryMessenger());
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
    }

    private void onAttachedToEngine(BinaryMessenger messenger) {
        channel = new MethodChannel(messenger, CHANNEL_NAME);
        channel.setMethodCallHandler(this);
    }


    @Override
    public void onMethodCall(MethodCall call, Result result) {
        switch (call.method) {
            case "login":
                List<String> permissions = call.argument("permissions");
                facebookAuth.login(this.activityPluginBinding.getActivity(), permissions, result);
                break;

            case "isLogged":
                facebookAuth.isLogged(result);
                break;

            case "getUserData":
                String fields = call.argument("fields");
                facebookAuth.getUserData(fields, result);
                break;

            case "logOut":
                facebookAuth.logOut(result);
                break;

            default:
                result.notImplemented();
        }

    }

    private void attachToActivity(ActivityPluginBinding binding) {
        this.activityPluginBinding = binding;
        activityPluginBinding.addActivityResultListener(facebookAuth.resultDelegate);
    }

    private void disposeActivity() {
        activityPluginBinding.removeActivityResultListener(facebookAuth.resultDelegate);
        // delegate.setActivity(null);
        activityPluginBinding = null;
    }

    @Override
    public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
        this.attachToActivity(binding);
    }

    @Override
    public void onDetachedFromActivityForConfigChanges() {
        disposeActivity();
    }

    @Override
    public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {
        this.attachToActivity(binding);
    }

    @Override
    public void onDetachedFromActivity() {
        disposeActivity();
    }
}


example/lib/main.dart
import 'dart:convert';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_facebook_auth/flutter_facebook_auth.dart';

void main() => runApp(MyApp());

String prettyPrint(Map json) {
  JsonEncoder encoder = new JsonEncoder.withIndent('  ');
  String pretty = encoder.convert(json);
  return pretty;
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  dynamic _userData;
  AccessToken _accessToken;
  bool _checking = true;

  String twitterStatus = "No Logged";

  @override
  void initState() {
    super.initState();
    _checkIfIsLogged();
  }

  _printCredentials(LoginResult result) {
    _accessToken = result.accessToken;
    print("userId: ${result.accessToken.userId}");
    print("token: ${prettyPrint(_accessToken.toJson())}");
    print("expires: ${result.accessToken.expires}");
    print("grantedPermission: ${result.accessToken.grantedPermissions}");
    print("declinedPermissions: ${result.accessToken.declinedPermissions}");
  }

  _checkIfIsLogged() async {
    final AccessToken accessToken = await FacebookAuth.instance.isLogged;
    setState(() {
      _checking = false;
    });
    if (accessToken != null) {
      print("is Logged:::: ${prettyPrint(accessToken.toJson())}");
      // now you can call to  FacebookAuth.instance.getUserData();
      final userData = await FacebookAuth.instance.getUserData();
      // final userData = await FacebookAuth.instance.getUserData(fields:"email,birthday");
      _accessToken = accessToken;
      setState(() {
        _userData = userData;
      });
    }
  }

  _login() async {
    setState(() {
      _checking = true;
    });
    final result = await FacebookAuth.instance.login();
    // final result = await FacebookAuth.instance.login(permissions:['email','user_birthday']);

    _checking = false;
    switch (result.status) {
      case FacebookAuthLoginResponse.ok:
        _printCredentials(result);
        // get the user data
        final userData = await FacebookAuth.instance.getUserData();
        // final userData = await _fb.getUserData(fields:"email,birthday");
        _userData = userData;
        break;
      case FacebookAuthLoginResponse.cancelled:
        print("login cancelled");
        break;
      default:
        print("login failed");
    }

    setState(() {});
  }

  _logOut() async {
    await FacebookAuth.instance.logOut();
    _accessToken = null;
    setState(() {
      _userData = null;
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Facebook Auth Example'),
        ),
        body: _checking
            ? Center(
                child: CircularProgressIndicator(),
              )
            : SingleChildScrollView(
                child: Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.center,
                    children: <Widget>[
                      Text(
                        _userData != null
                            ? prettyPrint(_userData)
                            : "NO LOGGED",
                      ),
                      SizedBox(height: 20),
                      _accessToken != null
                          ? Text(
                              prettyPrint(_accessToken.toJson()),
                            )
                          : Container(),
                      SizedBox(height: 20),
                      CupertinoButton(
                        color: Colors.blue,
                        child: Text(
                          _userData != null ? "LOGOUT" : "LOGIN",
                          style: TextStyle(color: Colors.white),
                        ),
                        onPressed: _userData != null ? _logOut : _login,
                      ),
                      SizedBox(height: 50),
                    ],
                  ),
                ),
              ),
      ),
    );
  }
}
